examples/string_token_substitution.ipynb

```diff
@@ -0,0 +1,247 @@
+   ],
+   "id": "feec93d227a1e95e"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "\n",
+    "## Example - LocatableContent\n",
+    "\n",
+    "Let's take an example from the MediaCreation library. The LocatableContent trait has an `\"isTemplated\"` property, which, if true, indicates that the `\"location\"` property contains substitution tokens.\n",
+    "\n",
+    "The `\"isTemplated\"` documentation says\n",
+    "\n",
+    "> The URL in the location property contains variables (post decoding) that must be expanded before any loading is attempted.\n",
+    ">\n",
+    "> Variables use the OpenAssetIO syntax (eg: \"image.{frame:04d}.{view}.exr\") see the OpenAssetIO documentation for more details.\n",
+    ">\n",
+    "> The following well-known variables are defined within the MediaCreation ecosystem, and must be used where applicable to any specific workflow:\n",
+    "> - frame: An integer frame number for the current time.\n",
+    "> - view: A string representing the current view (eg. \"left\").\n",
+    "\n",
+    "This implies that a file name of `\"image.{frame}.{view}.exr\"` should be resolved to `\"image.1.left.exr\"` if the frame number is 1 and the view is \"left\".\n",
+    "\n",
+    "\n"
+   ],
+   "id": "e1cf4ef1d84c6d3c"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "## The OpenAssetIO substitution syntax\n",
+    "\n",
+    "In general, the substitution tokens are of the form `{variable:format}` where `variable` is the name of the variable and `format` is a format string. The `format` part is optional. If included, the format specification is compatible with the C++ [libfmt](https://fmt.dev/latest/syntax.html) library, which in turn is broadly compatible with C++20's [`std::format`](https://en.cppreference.com/w/cpp/utility/format/format) and Python [format strings](https://docs.python.org/3.9/library/string.html#formatspec). \n",
+    "\n",
+    "For maximum cross language compatibility, only basic substitutions are officially supported. Specifically, substitutions with no format string, or integers with zero padding. This may be revised in the future.\n"
+   ],
+   "id": "f37a7c6450e02c50"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "## The `substitute` utility function\n",
+    "\n",
+    "The `substitute` utility function is a simple function that takes a string and a dictionary of substitutions, and returns a new string with the substitutions made. The dictionary should map variable names to their values.  Extra values will be ignored, and missing values is an error.\n",
```

Comment:
[Question] I think there's a decision hiding here that we need to make explicit. Ie. Is the `substitute` function _the way_ to do openassetio substitution, or is it merely a utility, and you're free to implement other substitution methods on your own.

This is sort of like an implementation hiding concern, based on whether we're going to change substitution rules later. I can imagine if we're not clear that "to do this properly, you _must_ use this function", then we could get complaints that we're doing a breaking change because hosts have to update their manual substitution code.

By framing this as "this is just a simple utility function" we sort of come down on the side that this is just here to help, but you can go ahead and do what you like. 

I think this is an error, as I don't think folks will really consider the interface agreement they're signing up for, and will just assume substitution behaviour will never change.

What do you think?

Reply:
`substitute` is just a utility function that hosts/managers don't _have_ to use - e.g. they can use `str.format(...)` in Python instead.

Changing "the OpenAssetIO format" would be a big breaking change - the specification of that is far more important than the `substitute` function. Managers _must_ supply strings that follow the format - the `subtstitute` function doesn't help with that, other than for testing.

I guess instead of saying you _must_ use this function, we should say hosts/managers _will_ use this function, so you'd better structure your strings to be compatible with it.

Reply:
Clarified in ad8acddb

Reply:
I don't really see a difference between `will` and `must`, they both imply that not using `substitute` is a programming error. 

The clarification is good, so i'm fine with the way things are.

As a P.S however, i'm not sure we fully understand eachother. I want to say that anyone "resolving" a string to its final form _must_ use substitute, as this minimises the potential are of breakage if we do change the standard. _Ideally_ there would be a "tokenize" method as well, to make changes in the standard non breaking in either direction. Even substitute as it stands probably isn't strong enough without the enhancements discussed in the TSC, as well as partial substitution, to properly put this edict down, but I hope one day it will become so.

Reply:
I wanted to be a bit weaker and allow use of `fmtlib`/`std::format`/Python format strings, with `substitute` just being a convenience.

Though indeed, enhancements to the format syntax as discussed in the TSC (https://github.com/OpenAssetIO/OpenAssetIO/issues/1322) will likely break that compatibility.